nameko/containers.py

```diff
@@ -353,41 +347,58 @@ def spawn_worker(self, entrypoint, args, kwargs,
-        return worker_ctx
+        gt = self._worker_pool.spawn(
+            self._run_worker, worker_ctx, handle_result
+        )
+        gt.link(self._handle_worker_thread_exited, worker_ctx)
 
-    def spawn_managed_thread(self, run_method, protected=False):
-        """ Spawn a managed thread to run ``run_method``.
+        self._worker_threads[worker_ctx] = gt
+        self.worker_created(worker_ctx)
+        return worker_ctx
 
-        Threads can be marked as ``protected``, which means the container will
-        not forcibly kill them until after all extensions have been killed.
-        Extensions that require a managed thread to complete their kill
-        procedure should ensure to mark them as ``protected``. For example,
-        :class:`nameko.messaging.QueueConsumer` cleanly closes connections
-        to the AMQP broker even when killed; the thread that holds that
-        connection is ``protected`` so it isn't stopped until after the
-        QueueConsumer returns from its kill.
+    def spawn_managed_thread(self, fn, protected=None, identifier=None):
+        """ Spawn a managed thread to run ``fn`` on behalf of an extension.
+        The passed `identifier` will be included in logs related to this
+        thread, and otherwise defaults to `fn.__name__`, if it is set.
 
-        Any uncaught errors inside ``run_method`` cause the container to be
-        killed.
+        Any uncaught errors inside ``fn`` cause the container to be killed.
 
         It is the caller's responsibility to terminate their spawned threads.
         Threads are killed automatically if they are still running after
         all extensions are stopped during :meth:`ServiceContainer.stop`.
 
         Extensions should delegate all thread spawning to the container.
         """
-        gt = eventlet.spawn(run_method)
-        if not protected:
-            self._active_threads[gt] = MANAGED_THREAD
-        else:
-            self._protected_threads.add(gt)
-        gt.link(self._handle_thread_exited)
+        if protected is not None:
+            warnings.warn(
+                "The signature of `spawn_managed_thread` has changed. "
+                "The `protected` kwarg is now deprecated, and extensions "
+                "can pass an idenifier as a keyword argument for better "
+                "logging. See :meth:`nameko.containers.ServiceContainer."
+                "spawn_managed_thread`.", DeprecationWarning
+            )
+        if identifier is None:
+            identifier = getattr(fn, '__name__', "<unknown>")
+
+        gt = eventlet.spawn(fn)
+        self._managed_threads[gt] = identifier
+        gt.link(self._handle_managed_thread_exited, identifier)
         return gt
 
+    def worker_created(self, worker_ctx):
+        """ Called after a worker is created.
+
+        This is a hook for `ServiceContainer` subclasses to track worker
+        creation.
+        """
+
+    def worker_destroyed(self, worker_ctx):
```

Comment:
That is the main use-case and I'm happy to make that change. I'll wait 'til you see my proposed implementation for the entrypoint waiter though, since there might be other things to discuss.